03_Modules/EVDriver/src/module/LocalAuthListService.ts

```diff
@@ -0,0 +1,124 @@
+// Copyright Contributors to the CitrineOS Project
+//
+// SPDX-License-Identifier: Apache 2.0
+
+import { AttributeEnumType, SendLocalListRequest, UpdateEnumType } from "@citrineos/base";
+import { IDeviceModelRepository, ILocalAuthListRepository, VariableCharacteristics, VariableAttribute, LocalListVersion, SendLocalList, } from "@citrineos/data";
+import { LocalListAuthorization } from "@citrineos/data/src/layers/sequelize/model/Authorization/LocalListAuthorization";
+
+import { v4 as uuidv4 } from 'uuid';
+
+export class LocalAuthListService {
+  protected _localAuthListRepository: ILocalAuthListRepository;
+  protected _deviceModelRepository: IDeviceModelRepository;
+
+  constructor(localAuthListRepository: ILocalAuthListRepository,
+    deviceModelRepository: IDeviceModelRepository
+  ) {
+    this._localAuthListRepository = localAuthListRepository;
+    this._deviceModelRepository = deviceModelRepository;
+  }
+
+  /**
+   * Validates a SendLocalListRequest and persists it, then returns the correlation Id.
+   *
+   * @param {string} stationId - The ID of the station to which the SendLocalListRequest belongs.
+   * @param {string} correlationId - The correlation Id that will be used for the SendLocalListRequest.
+   * @param {SendLocalListRequest} sendLocalListRequest - The SendLocalListRequest to validate and persist.
+   * @return {SendLocalList} The persisted SendLocalList.
+   */
+  async persistSendLocalListForStationIdAndCorrelationIdAndSendLocalListRequest(stationId: string, correlationId: string, sendLocalListRequest: SendLocalListRequest): Promise<SendLocalList> {
+    const localListVersion = await this._localAuthListRepository.readOnlyOneByQuery({
+      where: {
+        stationId: stationId,
+      },
+      include: [LocalListAuthorization],
+    });
+    const sendLocalList = await this.createSendLocalListFromStationIdAndRequestAndCurrentVersion(stationId, correlationId, sendLocalListRequest, localListVersion);
+
+    const newLocalAuthListLength = await this.countUpdatedAuthListFromRequestAndCurrentVersion(sendLocalList, localListVersion);
+    // TODO If Device Model is updated to allow different variable characteristics for the same variable per station, then we need to update this
+    const maxLocalAuthListEntries = await this.getMaxLocalAuthListEntries();
+    if (!maxLocalAuthListEntries) {
+      throw new Error('Could not get max local auth list entries, required by D01.FR.12');
+    } else if (newLocalAuthListLength > maxLocalAuthListEntries) {
+      throw new Error(`Updated local auth list length (${newLocalAuthListLength}) will exceed max local auth list entries (${maxLocalAuthListEntries})`);
```

Comment:
If this happens, don't we need to rollback? Seems we already persisted the new auth tokens, meaning our database is now lying about what is on the charger.

Reply:
No, the SendLocalList was created but the LocalListVersion remains unmodified. It only is updated when the SendLocalListResponse arrives.

Reply:
To clarify further, LocalListAuthorizations are made here, but they are only associated with the SendLocalList entry, which will later be used to update the actual LocalListVersion for the station when the SendLocalListResponse is processed.

There is still a concern for "orphaned" entries--SendLocalList entries that will never again be referenced. I don't see that concern (data consolidation) as being in scope for this module.